core/blockchain.go

```diff
@@ -811,6 +811,14 @@ func (bc *BlockChain) ExportN(w io.Writer, first uint64, last uint64) error {
 //
 // Note, this function assumes that the `mu` mutex is held!
 func (bc *BlockChain) writeHeadBlock(block *types.Block) {
+	// Normally the check at the end of the function will pass first, but if a node is restarted
+	// with the same l2-migration-block configured after already reaching and stopping on l2-migration-block,
+	// this check will pass first and log an error.
+	if bc.Config().IsL2Migration(block.Number()) {
+		log.Error("Attempt to insert block number >= l2MigrationBlock, stopping block insertion", "block", block.NumberU64(), "hash", block.Hash())
+		bc.StopInsert()
```

Comment:
I don't think we should need to call StopInsert here since this check will effectively prevent any additional blocks being added.

Reply:
See comment below on StopInsert()

Reply:
I wasn't sure why we need changes to `writeHeadBlock` at all. So I put a panic in for the `IsL2Migration` migration case to see when the check is relevant. The panic was triggered by the block fetcher.

It might be more elegant to discard the blocks when the block fetcher receives than rather then preventing the write here. Suggested commit: https://github.com/celo-org/celo-blockchain/pull/2330/commits/90dbc6cbd73b311f7472e190e1d025632f0e8f84

I still don't have the full picture, so let me know if I'm missing something.

Reply:
@karlb that looks pretty neat to me, and the tests seem to work, so I'd be in favour of this approach.

Reply:
Ok I've pushed a couple of updates to karlb/stop and removed basically everything except the extra flag, the rejection in the validate function and the e2e test. Tests seem to be working.